usage.c

```diff
@@ -6,17 +6,40 @@
 #include "git-compat-util.h"
 #include "cache.h"
 
+static void replace_control_chars(char *str, size_t size)
+{
+	size_t i;
+
+	for (i = 0; i < size; i++) {
+		if (iscntrl(str[i]) && str[i] != '\t' && str[i] != '\n')
+			str[i] = '?';
+	}
+}
+
+/* Atomically report (prefix + vsnprintf(err, params) + '\n') to stderr. */
 void vreportf(const char *prefix, const char *err, va_list params)
 {
 	char msg[4096];
-	char *p;
+	int ret;
+	size_t prefix_size, total_size;
 
-	vsnprintf(msg, sizeof(msg), err, params);
-	for (p = msg; *p; p++) {
-		if (iscntrl(*p) && *p != '\t' && *p != '\n')
-			*p = '?';
-	}
-	fprintf(stderr, "%s%s\n", prefix, msg);
+	prefix_size = strlen(prefix);
+	if (prefix_size >= sizeof(msg))
+		BUG("vreportf: prefix is too long");
+
+	memcpy(msg, prefix, prefix_size);
+
+	ret = vsnprintf(msg + prefix_size, sizeof(msg) - prefix_size, err, params);
+	if (ret < 0)
+		BUG("your vsnprintf is broken (returned %d)", ret);
```

Comment:
That'll cause an infinite loop, no?

Reply:
Not to my knowledge, because BUG() uses a different format string, which isn't broken.

Reply:
I'm currently testing this patch. Any feedback welcome, seriously. I do try to achieve high quality, and it takes some time.

Reply:
Do you agree?

Reply:
In my opinion, this is too fragile. If we're here, things are already going wrong, and I'd rather not add a recursion into the same function.

Reply:
Just exit silently?

Reply:
I would say, if `BUG()` is trustworthy, then it shouldn't be a problem to have 1 level of recursion. If `BUG()` is suspected to do very weird things, then it should rather be fixed.

Reply:
`BUG()` calls `vreportf()`. This recursion makes me uneasy. You might say that it cannot cause problems. But even you will have to admit that reasoning about this is not easy, and therein lies the opposite of perfect.

In any case, Git requires a `vsnprintf()` that returns the desired size, otherwise pretty much every `strbuf_addf()` call is broken (and there are many). So yes, if things go _so_ wrong as to let `vsnprintf()` return something negative, yes, I would exit silently.

And I would _certainly_ not introduce such a check in a commit that does many other things at the same time, too.

Reply:
I will change that to exiting silently.